### DISCUSSION TITLE
Issue with non symbolic MX matrices, inverse kinematics

### DISCUSSION THREAD

**Discussion by KarolineGrotterud** [casadi/casadi#3691] [Q&A]
Hi, 

I am trying to do inverse kinematics in casadi, but I have the issue that my MX matrices are not symbolic even though they consist of only symbolic parameters. Initializing a matrix as MX(lists) doesn't work, so I make all my matrices by writing 
matrix =  ca.MX.eye(4) and then swopping all the rows with a vertcat() like this:

```
Rx = ca.MX.eye(3)
    Rx[1,:] = ca.vertcat(0, ca.cos(psi), -ca.sin(psi))
    Rx[2,:] = ca.vertcat(0, ca.sin(psi), ca.cos(psi))
```

I have also tried this: 
```
    Rx = ca.MX.sym('Rx', 3,3)
    Rx[0,:] = ca.vertcat(1, 0, 0)
    Rx[1,:] = ca.vertcat(0, ca.cos(psi), -ca.sin(psi))
    Rx[2,:] = ca.vertcat(0, ca.sin(psi), ca.cos(psi))
```

Still I get Rx.is_symbolic() = False. 

I want to calculate the manipulability index of my robot like this
<img width="255" alt="image" src="https://github.com/casadi/casadi/assets/79454327/474aeda8-d981-4736-9362-8a0ce3c7c462">

 and I therefore need the output of my inverse kinematics function to be purely symbolic to use ca.jacobian().
How can I define my matrix expressions purely symbolic?

I have struggled with this for days, any help is much appreciated!

Here is a code snippet of my work. 

```
 def inverse_kinematics(rho, p_t, r_t, l):
        
    rho_1 = rho[0]
    rho_2 = rho[1]
    rho_3 = rho[2]

    x_t, y_t, z_t = p_t[0], p_t[1], p_t[2]
    psi_t = r_t[2]

    d_ap = 113

    x_p = x_t - rho_2 *ca.cos(rho_3+psi_t)+d_ap*ca.cos(rho_3+psi_t-rho_1)
    y_p = y_t - rho_2 *ca.sin(rho_3+psi_t)+d_ap*ca.sin(rho_3+psi_t-rho_1)
    phi_p = rho_3+psi_t-rho_1

    print("xp, yp: ", x_p, y_p)
    #print("phi_p: ", phi_p[0], "type: ", type(ca.vertcat(0,0,phi_p)))
    print("list: ", ca.vertcat(0,0,phi_p))
    
    # R_sp = sc.Rotation.as_matrix(sc.Rotation.from_euler("z", phi_p))
    R_sp = euler_rotation_matrix(0,0,phi_p)
    print("R_sp is symbolic? ", R_sp.is_symbolic())

    T_sp = ca.MX.eye(4)
    T_sp[:3, :3] = R_sp
    T_sp[:3, 3] = ca.vertcat(x_p, y_p, 0)
    print(T_sp)
    T_pa = ca.MX.eye(4)
    T_pa[0,:] = ca.vertcat(1, 0, 0, -d_ap)
    T_pa[1,:] = ca.vertcat(0, 1, 0, 0)
    T_pa[2,:] = ca.vertcat(0, 0, 1, 265.5)
                        
    T_sa = T_sp @ ca.MX(T_pa)
    T_d = ca.MX.eye(4)
    print("Is T_d symbolic? ", T_d.is_symbolic())
    #T_d[:3, :3] = sc.Rotation.as_matrix(sc.Rotation.from_euler("xyz", r_t))
    T_d[:3, :3] = euler_rotation_matrix(r_t[0], r_t[1], r_t[2])
    print("r_t symbolic?: ", r_t.is_symbolic())
    T_d[:3, 3] = ca.vertcat(x_t, y_t, z_t)
    T_at = ca.inv(T_sa) @ T_d

    joint_angles_arr = analytic_inverse_kinematics(T_at, l)

    return joint_angles_arr 
```

**Comment by tmmsartor** [comment]
The jacobian can be calculate with the respect of a purely symbolic expression, like:
`x = SX.sym("x")` or `y = MX.sym("y",2,2)`, etc.
As soon as you use any operator, the expression is not purely symbolic, i.e. `2*x` or `x+x`, etc, those are symbolic expression, but they are not purely symbolic because they contain operations. 

If you want to compute the derivative of some expression dependent on $P_{ee}$, with the respect of $P_{ee}$ and you can define 
If  $P_{ee}$ is a 4 by 4 matrix one can define `Pee = MX.sym("Pee",4,4)`  (can also be a sparse matrix), and then call `jacobian(<expression>,Pee)`

I am not sure if this is what you need, maybe you can provide more context around your manipulability index formula.
